python/tank/util/shotgun.py

```diff
@@ -1157,117 +1233,151 @@ def _create_published_file(tk, context, path, name, version_number, task, commen
+            # # than relying on Shotgun to compute this
+            # supports_specific_storage_syntax = (
+            #     hasattr(tk.shotgun, "server_caps") and
+            #     tk.shotgun.server_caps.version and
+            #     tk.shotgun.server_caps.version >= (6, 3, 17)
+            # )
+            #
+            # if supports_specific_storage_syntax:
+            #     # explicitly pass relative path and storage to shotgun
+            #     storage = tk.shotgun.find_one("LocalStorage", [["code", "is", storage_name]])
+            #
+            #     if storage is None:
+            #         # there is no storage in Shotgun that matches the one toolkit expects.
+            #         # this *may* be ok because there may be another storage in Shotgun that
+            #         # magically picks up the publishes and associates with them. In this case,
+            #         # issue a warning and fall back on the server-side functionality
+            #         log.warning(
+            #             "Could not find the expected storage '%s' in Shotgun to associate "
+            #             "publish '%s' with - falling back to Shotgun's built-in storage "
+            #             "resolution logic. It is recommended that you add the '%s' storage "
+            #             "to Shotgun" % (storage_name, path, storage_name))
+            #         data["path"] = {"local_path": path}
+            #
+            #     else:
+            #         data["path"] = {"relative_path": path_cache, "local_storage": storage}
+            #
+            # else:
+            #     # use previous syntax where we pass the whole path to Shotgun
+            #     # and shotgun will do the storage/relative path split server side.
+            #     # This operation may do unexpected things if you have multiple
+            #     # storages that are identical or overlapping
+            #     data["path"] = {"local_path": path}
+
+            # fill in the path cache field which is used for filtering in Shotgun
+            # (because SG does not support
+            data["path_cache"] = path_cache
 
-    if published_file_type:
-        if published_file_entity_type == "PublishedFile":
-            data["published_file_type"] = published_file_type
-        else: # == TankPublishedFile
-            data["tank_type"] = published_file_type
+        else:
+
+            # path does not map to any configured root - fall back gracefully:
+            # 1. look for storages in Shotgun and see if we can create a local path
+            # 2. failing that, just register the entry as a file:// resource.
+            log.debug("Path '%s' does not have an associated config root." % path)
+            log.debug("Will check shotgun local storages to see if there is a match.")
+
+            matching_local_storage = False
+            log.debug("Retrieving local storages from Shotgun...")
+            storages = tk.shotgun.find(
```

Comment:
agreed - my main concern here was to make sure the existing behaviour wasn't slowed down. I'll make a polish ticket for speeding up publishing.